builder/builder.go

```diff
@@ -412,97 +427,454 @@ func (b *Builder) getUpstreamBundles(ver string, prune bool) error {
+	}
+
+	// Assign "top level" bundles
+	switch listType {
+	case MixList:
+		bundles = mixBundles
+	case LocalList:
+		bundles = localBundles
+	case UpstreamList:
+		// Lazy-loaded because it's only needed for UpstreamList's
+		upstreamBundles, err = b.getDirBundlesListAsSet(getUpstreamBundlesPath(b.UpstreamVer))
+		if err != nil {
+			return err
+		}
+		bundles = upstreamBundles
+	}
+
+	// Used for pretty-printing lists
+	var max int
+	for _, bundle := range bundles {
+		if len(bundle.Name) > max {
+			max = len(bundle.Name)
+		}
+	}
+
+	// Create the full, parsed set
+	set, err := b.getFullBundleSet(bundles)
+	if err != nil {
+		return err
+	}
+
+	if tree {
+		// Print the tree view
+		sorted := getBundleSetKeysSorted(bundles)
+		for _, bundle := range sorted {
+			b.bundleTreePrint(set, bundle, 0, make([]bool, 0))
+		}
+
+		return nil
+	}
+
+	// Print a flat list and return
+	switch listType {
+	case MixList:
+		// Print the full, parsed set
+		sorted := getBundleSetKeysSorted(set)
+		for _, bundle := range sorted {
+			val := fmt.Sprintf("%-*s", max, set[bundle].Name)
+
+			if _, exists := localBundles[set[bundle].Name]; exists {
+				val += " (local)   "
+			} else {
+				val += " (upstream)"
+			}
+
+			if _, exists := bundles[set[bundle].Name]; exists {
+				val += " (included)"
+			}
+
+			fmt.Println(val)
+		}
+	case LocalList:
+		// Only print the top-level set
+		sorted := getBundleSetKeysSorted(bundles)
+		for _, bundle := range sorted {
+			val := fmt.Sprintf("%-*s", max, bundle)
+
+			if _, exists := mixBundles[bundle]; exists {
+				val += " (in mix)"
+			}
+
+			fmt.Println(val)
+		}
+	case UpstreamList:
+		// Only print the top-level set
+		sorted := getBundleSetKeysSorted(bundles)
+		for _, bundle := range sorted {
+			val := fmt.Sprintf("%-*s", max, bundle)
+
+			if _, exists := mixBundles[bundle]; exists {
+				val += " (in mix)"
+			} else {
+				val += "         "
+			}
+			if _, exists := localBundles[bundle]; exists {
+				val += " (masked by local)"
+			}
+
+			fmt.Println(val)
+		}
+	}
+
+	return nil
 }
 
 // UpdateMixVer automatically bumps the mixversion file +10 to prepare for the next build
 // without requiring user intervention. This makes the flow slightly more automatable.
 func (b *Builder) UpdateMixVer() error {
-	mixver, _ := strconv.Atoi(b.MixVer)
-	return ioutil.WriteFile(filepath.Join(b.VersionDir, "mixversion"), []byte(strconv.Itoa(mixver+10)), 0644)
+	mixVer, _ := strconv.Atoi(b.MixVer)
+	return ioutil.WriteFile(filepath.Join(b.VersionDir, "mixversion"), []byte(strconv.Itoa(mixVer+10)), 0644)
+}
+
+// createMixBundleDir generates the mix-bundles/ dir for chroot building. It will
+// clear the dir if it exists, compute the full list of bundles for the mix, and
+// copy the corresponding bundle files into mix-bundles/
+// Note: this function is only needed for the old Bundle Chroot Builder, and can
+// be removed once the UseNewChrootBuilder flag is gone.
+func (b *Builder) createMixBundleDir() error {
```

Comment:
See other comments about creating the full mix once, if you take that route, you might just get a set here as parameter.

Reply:
I really wouldn't worry about optimizing this at all, as it's temporary code that will go away completely once we switch fully to new BCB.

Or I'm still misunderstanding what you mean, which is entirely possible :)

Reply:
I'd still probably do this one. (Not really worried about optimization in this case).
To make it clear (hah):

- in build chroots, get the set before both codepaths start
- validate the set (why not?)
- then we have a if for new and old, for the old, you call this function passing the set.

Reply:
So, with your approval, I'm going to skip this.  Yes, right now, each path in the code does the same thing, and yes, that could absolutely be pulled out above the two. But a) one of those paths is going away, and far more importantly b) trying to pull it up was way more annoying than I wanted to deal with. Pulling it up means `err` is already in scope, which makes the lint throw up a bunch of complaints about shaddowing definitions, even though it builds. I tried going through and changing all of the `:=`'s to `=`'s where it asked me to, but then it wanted me to move declarations of `outfile`, and it was just turning into a "if you give a mouse a cookie" situation, snowballing.

Reply:
Fine. Vetshadowing and :='s can get in the way of moving code around, indeed.